src/fdc/D88DiskImage.cc

```diff
@@ -0,0 +1,350 @@
+};
+
+[[nodiscard]] static /*constexpr*/ bool isValidD88Header(const D88Header& header, const std::array<int32_t, 164>& trackOffsets)
+{
+	if (header.writeProtected != one_of(0x00, 0x10)) {
+		return false;
+	}
+	if (header.diskType != one_of(D88_DISK_TYPE_2DD, D88_DISK_TYPE_1DD)) {
+		return false;
+	}
+
+    // check track offset
+    uint32_t prevTrackOffs = trackOffsets[0];
+	if (prevTrackOffs != sizeof(D88Header) + 4 * D88_TRACK_MAX)
+		return false;
+    for(int32_t i=1; i<160; ++i)
+    {
+        auto trackOffs = trackOffsets[i];
+        auto trackSize = trackOffs - prevTrackOffs;
+        if (trackSize <= 128 || trackSize >= 0x4000)
+            return false;
+		prevTrackOffs = trackOffs;
+    }
+    for(int32_t i=160; i<D88_TRACK_MAX; ++i)
+    {
+        if (trackOffsets[i] != 0)
+            return false;
+    }
+
+	return std::ranges::all_of(header.reserved, [](auto& r) { return r == 0; });
+}
+
+D88DiskImage::D88DiskImage(const Filename& filename, std::shared_ptr<File> file_)
+    : Disk(DiskName(std::move(filename)))
+    , file(std::move(file_))
+	, trackData(8192)
+	, cachedTrackNo(-1)
+{
+    D88Header header;
+    file->seek(0);
+    file->read(std::span{&header, 1});
+    file->read(std::span{&trackOffsets, 1});
+	if (!isValidD88Header(header, trackOffsets)) {
+		throw MSXException("Not a D88 image");
+	}
+
+    diskSize = header.diskSize;
+    singleSided = header.diskType == D88_DISK_TYPE_1DD;
+    writeProtected = header.writeProtected == 0x10;
+
+	// read track 0 first sector
+	D88SectorHeader sectorHeader;
+	file->seek(trackOffsets[0]);
+	file->read(std::span{ &sectorHeader, 1 });
+	sectorsPerTrack = sectorHeader.sectorsPerTrack;
+
+	trackSectors.reserve(32);
+}
+
+bool D88DiskImage::readD88Track(uint8_t track, uint8_t side)
+{
+    int32_t track_ = singleSided ? track : track * 2 + side;
+
+    if (track_ >= 160)
+        return false;
+
+	// check readed track.
+	if (cachedTrackNo == track_)
+		return true;
+	cachedTrackNo = -1;
+
+    auto trackOffs = trackOffsets[track_];
+    auto nextTrackOffs = trackOffsets[track_ + 1];
+    if (nextTrackOffs < 1)
+        nextTrackOffs = diskSize;
+    auto trackSize = nextTrackOffs - trackOffs;
+
+    if (trackOffs==0 || trackSize<1)
+        return false;
+
+	if (trackData.size() < trackSize)
+	{
+		trackData.resize(trackSize);
+	}
+	trackSectors.clear();
+
+	/* Read TrackData */
+    file->seek(trackOffs);
+	file->read(std::span{ trackData });
+
+    uint16_t inTrackOffset = 0;
+    while(inTrackOffset < trackSize)
+    {
+		D88SectorHeader* pHeader = (D88SectorHeader*)(trackData.data() + inTrackOffset);
+		D88SectorData sector;
+
+		int nn = pHeader->n & 3;
+		sector.c = pHeader->c;
+		sector.h = pHeader->h;
+		sector.r = pHeader->r;
+		sector.n = pHeader->n;
+		sector.deleteFlag = pHeader->deleteFlag;
+		sector.status = pHeader->status;
+		sector.sectorSize = 128 << nn;
+		sector.dataOffset = inTrackOffset + 16;
+		sector.dataSize = pHeader->sectorDataSize;

```

Comment:
And here we should also protect better:
* It could be that `inTrackOffset + 16 + sector.dataSize > trackSize`.
* If that's the case then later in `readTrack()`, `trackDataSpan.subspan(sct.dataOffset, sct.dataSize)` will be past the end of the `trackData` buffer.

--

The [D88 documentation](https://www.pc98.org/project/doc/d88.html) suggests to NOT use the `sectorDataSize` field (because too many D88 images have a corrupt value for that field). Instead it should be calculated like you do above for `sectorSize`.